src/wp-includes/user.php

```diff
@@ -3012,19 +3012,57 @@ function retrieve_password( $user_login = null ) {
 	 */
 	$message = apply_filters( 'retrieve_password_message', $message, $key, $user_login, $user_data );
 
+	// Wrap the single notification email arguments in an array to pass them to the retrieve_password_notification_email filter.
+	$retrieve_password_notification_email = array(
+		'to'      => $user_email,
+		'subject' => $title,
+		'message' => $message,
+		'headers' => '',
+	);
+
+	/**
+	 * Filters the contents of the reset password notification email sent to the user.
+	 *
+	 * @since x.x.x
+	 *
+	 * @param array   $retrieve_password_notification_email {
+	 *     Used to build wp_mail().
+	 *
+	 *     @type string $to      The intended recipient - user email address.
+	 *     @type string $subject The subject of the email.
+	 *     @type string $message The body of the email.
+	 *     @type string $headers The headers of the email.
+	 * }
+	 * @param string  $key        The activation key.
+	 * @param string  $user_login The username for the user.
+	 * @param WP_User $user_data  WP_User object.
+	 */
+	$retrieve_password_notification_email = apply_filters( 'retrieve_password_notification_email', $retrieve_password_notification_email, $key, $user_login, $user_data );
+
 	if ( $switched_locale ) {
 		restore_previous_locale();
 	}
 
-	if ( $message && ! wp_mail( $user_email, wp_specialchars_decode( $title ), $message ) ) {
-		$errors->add(
-			'retrieve_password_email_failure',
-			sprintf(
-				/* translators: %s: Documentation URL. */
-				__( '<strong>Error</strong>: The email could not be sent. Your site may not be correctly configured to send emails. <a href="%s">Get support for resetting your password</a>.' ),
-				esc_url( __( 'https://wordpress.org/support/article/resetting-your-password/' ) )
-			)
+	if ( $message ) {
+
+		$wp_mail_result = wp_mail(
+			$retrieve_password_notification_email['to'],
+			wp_specialchars_decode( $retrieve_password_notification_email['subject'] ),
+			$retrieve_password_notification_email['message'],
+			$retrieve_password_notification_email['headers']
 		);
```

Comment:
Think this is great, but then we should also remove the `empty()` check for `message` from
```php
if ( is_array( $notification_email ) && ! empty( $notification_email['message'] ) ) {
```

Reply:
In this case, `empty()` [will not error](https://3v4l.org/qEcch) if the `message` key is missing. Currently, if the `message` fails a falsey check, the email is not sent. To change this now would change the **behaviour** of this function.

However, this means that we've had emails go silently unsent up until now. That's a bad outcome, particularly for password reset emails. I agree that we should still remove this `! empty()` check, as PHPMailer [already validates](https://github.com/PHPMailer/PHPMailer/blob/8c952fb573d3ab837cf9780257e0605be53afffc/src/PHPMailer.php#L1578) empty `message` values and, to my knowledge, we don't override the `allowEmpty` value.

Nice catch @MarloKessler 👍

Reply:
I had another thought about this and recoded some stuff. Now, if an extender returns `false` no password reset mail is sent. If the extender returns an array, the array is merged with defaults as you described above. If the filter does not return an array, the defaults are used. Thus, we give extenders the option to prevent sending reset mails (*Is that what we want?*) and also fallback on defaults as you pointed out above. Have a look at the code: https://github.com/WordPress/wordpress-develop/pull/2082/files#diff-ff818ec19f0f94e4175f72bd2cc4cd67c455c0c50e5b651690a1e312c284ba06R3053-R3077

Reply:
I agree that we should default to `$defaults` when `$notification_email` isn't an array. 👍 

I'm not sure if there's a use case for when we'd want to allow blocking of password reset emails. The `$message` falsey check was implemented from [the beginning ](https://core.trac.wordpress.org/attachment/ticket/34281/34281.diff) to guard against a falsey `$message` value after the `retrieve_password_message` filter.

However, PHPMailer already defaults to `PHPMailer->AllowEmpty = false` and [throws an Exception](https://github.com/PHPMailer/PHPMailer/blob/d4bf3504b93c38c7bfebc9c686471f48e6f84c06/src/PHPMailer.php#L1580) for "message body empty". The `$message` falsey check prevented PHPMailer from validating this and the email was silently unsent.

However, it may be that I just can't think of a use case. Can you think of any?

cc'ing @audrasjb and @adamsilverstein for their input on this part in case they can think of any use cases where we want to allow the password reset emails to be unsent without any output.

Reply:
Mhh, I currently can't think of any case.

Reply:
While it's good to facilitate possible use cases, once they're in, removing them becomes a potential BC break. In that sense, it's _sometimes_ better to wait for the enhancement to be requested so that we know it's useful, rather than allowing potentially accidental `false` values to stop a password reset email from being sent. 

Without this check, `$notification_email` will fail the `is_array()` check and will then become equal to `$defaults`, meaning that in the event of a developer error, users can still get their password reset email - albeit, with different content/styling than the developer intended.

I'd say let's put a pin in this part of the PR _just for now_ then - others will be back soon after the holidays and may have some ideas. There's still unit tests to add to the PR and possibly some validation, if necessary.

Reply:
Okay sounds good! I removed the check

Reply:
I also added a unit test for the `retrieve_password` function. Since I'm not that familiar with WP unit tests it would be great if you could have a look on this.

Reply:
Will do! Quick tip if you weren't aware: `vendor/bin/phpcbf <relative/path/to/file>` will do some PHPCS autofixes for you.

Reply:
Thank you!